File: examples/extensions/python-conformance/full_conformance_extension.py
from __future__ import annotations

from typing import Any

from noloong_jsonrpc import ExtensionContext, JsonDict, serve_extension

MODEL_ID = "conformance-model"
TOOL_NAME = "conformance_echo"
CONTEXT_ID = "conformance-context"
PHASE_ID = "conformance.phase"
PHASE_HOOK_ID = "conformance-hook"
TOOL_HOOK_ID = "conformance-tool-hook"
COMPACTION_ID = "conformance-compaction"


def initialize(params: JsonDict, _context: ExtensionContext) -> JsonDict:
    if params.get("protocolVersion") != 1:
        raise ValueError("unsupported protocol version")
    return {"manifest": {"name": "python-conformance-extension", "version": "0.1.0"}}


def list_capabilities(_params: JsonDict, _context: ExtensionContext) -> JsonDict:
    return {
        "capabilities": [
            {"type": "model_provider", "id": MODEL_ID},
            {
                "type": "tool",
                "spec": {
                    "name": TOOL_NAME,
                    "description": "Echo text from the Python conformance extension",
                    "inputSchema": {
                        "type": "object",
                        "properties": {"text": {"type": "string"}},
                        "required": ["text"],
                    },
                    "permissions": [
                        {
                            "capability": "conformance.echo",
                            "description": "Allows the conformance echo tool to run",
                            "metadata": {"example": "python"},
                        }
                    ],
                },
            },
            {"type": "context_provider", "id": CONTEXT_ID},
            {"type": "phase_node", "id": PHASE_ID},
            {"type": "phase_hook", "id": PHASE_HOOK_ID},
            {"type": "tool_call_hook", "id": TOOL_HOOK_ID},
            {"type": "compaction_summarizer", "id": COMPACTION_ID},
        ]
    }


def stream_model(params: JsonDict, context: ExtensionContext) -> JsonDict:
    stream_id = required_string(params.get("streamId"), "streamId")
    request = record(params.get("request"))
    messages = request.get("messages")

    if not has_tool_result(messages):
        context.stream_event(
            stream_id,
            {
                "type": "tool_call",
                "tool_call": {
                    "id": "conformance-call-1",
                    "name": TOOL_NAME,
                    "arguments": {"text": "from model"},
                },
            },
        )
        context.stream_event(stream_id, finish_event("tool_use"))
        return {"streamId": stream_id}

    context.stream_event(stream_id, {"type": "started", "stream_id": stream_id})
    context.stream_event(stream_id, {"type": "text_delta", "text": "adapter complete"})
    context.stream_event(stream_id, finish_event("stop"))
    return {"streamId": stream_id}


def execute_tool(params: JsonDict, _context: ExtensionContext) -> JsonDict:
    request = record(params.get("request"))
    arguments = record(request.get("arguments"))
    return {
        "content": [{"type": "text", "text": str(arguments.get("text", ""))}],
        "details": {"example": "python"},
        "isError": False,
        "updates": [
            {
                "content": [{"type": "text", "text": "tool update"}],
                "details": {"step": 1},
            }
        ],
    }


def apply_context(_params: JsonDict, _context: ExtensionContext) -> JsonDict:
    return {
        "effects": [patch_context_effect("conformance_context", True)]
    }


def run_phase(params: JsonDict, _context: ExtensionContext) -> JsonDict:
    request = record(params.get("request"))
    return {
        "scratch": record(request.get("scratch")),
        "effects": [patch_context_effect("conformance_phase", True)],
        "streamEvents": [],
        "resolvedToolCalls": [],
        "toolOutputs": [],
    }


def run_phase_hook(params: JsonDict, _context: ExtensionContext) -> JsonDict:
    require_hook(params, PHASE_HOOK_ID)
    return {}


def run_tool_hook(params: JsonDict, _context: ExtensionContext) -> JsonDict:
    require_hook(params, TOOL_HOOK_ID)
    if params.get("hookPoint") == "before_tool_call":
        return {
            "decision": {
                "outcome": "allow",
                "reason": "allowed by Python conformance tool hook",
                "approver": "python-conformance-extension",
                "metadata": {"example": "python"},
            }
        }
    return {}


def summarize_compaction(params: JsonDict, _context: ExtensionContext) -> JsonDict:
    messages = params.get("messagesToSummarize")
    count = len(messages) if isinstance(messages, list) else 0
    return {
        "summary": f"conformance compaction summary: {count}",
        "metadata": {"example": "python"},
    }


def shutdown(_params: JsonDict, _context: ExtensionContext) -> JsonDict:
    return {}


def finish_event(stop_reason: str) -> JsonDict:
    return {"type": "finished", "stop_reason": stop_reason}


def has_tool_result(messages: Any) -> bool:
    return isinstance(messages, list) and any(record(message).get("role") == "tool_result" for message in messages)


def require_hook(params: JsonDict, expected_id: str) -> None:
    if params.get("hookId") != expected_id:
        raise ValueError(f"unexpected hook id: {params.get('hookId')}")


def required_string(value: Any, field: str) -> str:
    if not isinstance(value, str):
        raise ValueError(f"missing string field: {field}")
    return value


def record(value: Any) -> JsonDict:
    return value if isinstance(value, dict) else {}


def patch_context_effect(key: str, value: Any) -> JsonDict:
    return {
        "type": "patch_context",
        "patch": {"op": "set", "key": key, "value": value},
    }


if __name__ == "__main__":
    serve_extension(
        {
            "initialize": initialize,
            "capabilities/list": list_capabilities,
            "model/stream": stream_model,
            "tool/execute": execute_tool,
            "context/apply": apply_context,
            "phase/run": run_phase,
            "phase_hook/run": run_phase_hook,
            "tool_hook/run": run_tool_hook,
            "compaction/summarize": summarize_compaction,
            "shutdown": shutdown,
        }
    )

File: examples/extensions/typescript-conformance/src/full-conformance-extension.ts
import {
  createStdioJsonRpcExtension,
  type ExtensionContext,
  type JsonObject,
  type JsonValue,
} from "./noloong.ts";

const MODEL_ID = "conformance-model";
const TOOL_NAME = "conformance_echo";
const CONTEXT_ID = "conformance-context";
const PHASE_ID = "conformance.phase";
const PHASE_HOOK_ID = "conformance-hook";
const TOOL_HOOK_ID = "conformance-tool-hook";
const COMPACTION_ID = "conformance-compaction";

createStdioJsonRpcExtension({
  initialize(params) {
    if (params.protocolVersion !== 1) {
      throw new Error("unsupported protocol version");
    }
    return {
      manifest: {
        name: "typescript-conformance-extension",
        version: "0.1.0",
      },
    };
  },

  "capabilities/list"() {
    return {
      capabilities: [
        { type: "model_provider", id: MODEL_ID },
        {
          type: "tool",
          spec: {
            name: TOOL_NAME,
            description: "Echo text from the TypeScript conformance extension",
            inputSchema: {
              type: "object",
              properties: {
                text: { type: "string" },
              },
              required: ["text"],
            },
            permissions: [
              {
                capability: "conformance.echo",
                description: "Allows the conformance echo tool to run",
                metadata: { example: "typescript" },
              },
            ],
          },
        },
        { type: "context_provider", id: CONTEXT_ID },
        { type: "phase_node", id: PHASE_ID },
        { type: "phase_hook", id: PHASE_HOOK_ID },
        { type: "tool_call_hook", id: TOOL_HOOK_ID },
        { type: "compaction_summarizer", id: COMPACTION_ID },
      ],
    };
  },

  "model/stream"(params, context) {
    const streamId = requiredString(params.streamId, "streamId");
    const request = objectValue(params.request);
    const messages = request.messages;

    if (!hasToolResult(messages)) {
      context.streamEvent(streamId, {
        type: "tool_call",
        tool_call: {
          id: "conformance-call-1",
          name: TOOL_NAME,
          arguments: { text: "from model" },
        },
      });
      context.streamEvent(streamId, finishEvent("tool_use"));
      return { streamId };
    }

    context.streamEvent(streamId, { type: "started", stream_id: streamId });
    context.streamEvent(streamId, { type: "text_delta", text: "adapter complete" });
    context.streamEvent(streamId, finishEvent("stop"));
    return { streamId };
  },

  "tool/execute"(params) {
    const request = objectValue(params.request);
    const args = objectValue(request.arguments);
    return {
      content: [{ type: "text", text: String(args.text ?? "") }],
      details: { example: "typescript" },
      isError: false,
      updates: [
        {
          content: [{ type: "text", text: "tool update" }],
          details: { step: 1 },
        },
      ],
    };
  },

  "context/apply"() {
    return {
      effects: [patchContextEffect("conformance_context", true)],
    };
  },

  "phase/run"(params) {
    const request = objectValue(params.request);
    return {
      scratch: objectValue(request.scratch),
      effects: [patchContextEffect("conformance_phase", true)],
      streamEvents: [],
      resolvedToolCalls: [],
      toolOutputs: [],
    };
  },

  "phase_hook/run"(params) {
    requireHook(params, PHASE_HOOK_ID);
    return {};
  },

  "tool_hook/run"(params) {
    requireHook(params, TOOL_HOOK_ID);
    if (params.hookPoint === "before_tool_call") {
      if (stateHasUserText(params, "approval")) {
        return {
          approval: {
            prompt: "Approve TypeScript conformance tool?",
            reason: "TypeScript conformance approval case",
            metadata: { example: "typescript" },
          },
        };
      }
      return {
        decision: {
          outcome: "allow",
          reason: "allowed by TypeScript conformance tool hook",
          approver: "typescript-conformance-extension",
          metadata: { example: "typescript" },
        },
      };
    }
    return {};
  },

  "compaction/summarize"(params) {
    const messages = Array.isArray(params.messagesToSummarize)
      ? params.messagesToSummarize
      : [];
    return {
      summary: `conformance compaction summary: ${messages.length}`,
      metadata: { example: "typescript" },
    };
  },

  shutdown() {
    return {};
  },
}).serve().catch((err: unknown) => {
  process.stderr.write(`${err instanceof Error ? err.message : String(err)}\n`);
  process.exitCode = 1;
});

function finishEvent(stopReason: "stop" | "tool_use"): JsonObject {
  return { type: "finished", stop_reason: stopReason };
}

function hasToolResult(messages: unknown): boolean {
  return Array.isArray(messages) && messages.some((message) => objectValue(message).role === "tool_result");
}

function stateHasUserText(params: JsonObject, expected: string): boolean {
  const state = objectValue(params.state);
  const messages = state.messages;
  if (!Array.isArray(messages)) {
    return false;
  }
  return messages.some((messageValue) => {
    const message = objectValue(messageValue);
    if (message.role !== "user" || !Array.isArray(message.content)) {
      return false;
    }
    return message.content.some((blockValue) => {
      const block = objectValue(blockValue);
      return block.type === "text" && block.text === expected;
    });
  });
}

function requireHook(params: JsonObject, expectedId: string) {
  const hookId = params.hookId;
  if (hookId !== expectedId) {
    throw new Error(`unexpected hook id: ${String(hookId)}`);
  }
}

function requiredString(value: JsonValue | undefined, field: string): string {
  if (typeof value !== "string") {
    throw new Error(`missing string field: ${field}`);
  }
  return value;
}

function objectValue(value: unknown): JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as JsonObject)
    : {};
}

function patchContextEffect(key: string, value: JsonValue): JsonObject {
  return {
    type: "patch_context",
    patch: { op: "set", key, value },
  };
}
